---
title: "Class 12: RNASeq analysis"
author: "Jennifer Thai (A17893762)"
format: pdf
toc: TRUE
---

## Background

Today we will analyze some RNASeq data from HImes et al. on the effects of a common steroid (dexamethasone) on airway smooth muscle cells (ASM cells).

Our starting point is the "counts" data and the "metadata" that contain the count values for each genes in their different experiments (i.e. cell lines with or without the drug).

## Data Import 

```{r}
# Complete the missing code
counts <- read.csv("airway_scaledcounts.csv", row.names=1)
metadata <-  read.csv("airway_metadata.csv")
```

Let's have a wee peak at these objects:

```{r}
head(counts)
```

> Q1. How many genes are in this dataset?

```{r}
nrow(counts)
```

```{r}
metadata
```

> Q. How many different experiments (columns in counts or rows in metadata) are there?

```{r}
ncol(counts)
```
```{r}
nrow(metadata)
```

> Q2. How many ‘control’ cell lines do we have?

```{r}
sum(metadata$dex == "control")
```

## Toy differential gene expression

To start our analysis, let's calculate the mean counts for all genes in the "control" experiments.

1. Extract all "control" columns from the `counts` object
2. Calculate the mean for all rows (i.e. genes) of these "control" columns

3-4. Do the same for "treated"
5. Compare these `control.mean` and `treated.mean` values

> Q3. How would you make the above code in either approach more robust? Is there a function that could help here?

```{r}
control.inds <- metadata$dex == "control"
control.counts <- counts[ , control.inds]
```

```{r}
control.means <- rowMeans(control.counts)
```

> Q4. Follow the same procedure for the treated samples (i.e. calculate the mean per gene across drug treated samples and assign to a labeled vector called treated.mean)

```{r}
treated.inds <- metadata$dex == "treated"
treated.counts <- counts[ , treated.inds]
```

```{r}
treated.means <- rowMeans(treated.counts)
```

Store these together for ease of bookkeeping as `meancounts`

```{r}
meancounts <- data.frame(control.means, treated.means)
head(meancounts)
```

> Q5 (a). Create a scatter plot showing the mean of the treated samples against the mean of the control samples. Your plot should look something like the following.

Make a plot of control vs treated mean values for all genes

```{r}
plot(meancounts)
```

> Q5 (b).You could also use the ggplot2 package to make this figure producing the plot below. What geom_?() function would you use for this plot?

```{r}
library(ggplot2)

ggplot(meancounts, aes(control.means, treated.means)) +
  geom_point()
```

> Q6. Try plotting both axes on a log scale. What is the argument to plot() that allows you to do this?

Make this a log plot

```{r}
plot(meancounts, log="xy")
```

We often talk about metrics like "log2 fold-change"

```{r}
# treated/control
log2(10/10)
```

```{r}
log2(10/20)
```
```{r}
log2(20/10)
```

```{r}
log2(40/10)
log2(10/40)
```

Let's calculate the log2 fold change for our treated over control counts.

```{r}
meancounts$log2fc <- 
log2( meancounts$treated.means / meancounts$control.means )
```

```{r}
head(meancounts)
```

```{r}
zero.vals <- which(meancounts[ , 1:2]==0, arr.ind=TRUE)

to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm, ]
head(mycounts)
```

> Q7. What is the purpose of the arr.ind argument in the which() function call above? Why would we then take the first column of the output and need to call the unique() function?

The argument `arr.ind()` in the which() function tells R to give the vector position of the TRUE values. The unique() function will return only 1 of that value instead of giving duplicates. 


A common "rule of thumb" is a log2 fold change has a cutoff of +2 and -2 to call genes "Up regulated" or "Down regulated".

> Q8. Determine how many up regulated genes we have at the greater than 2 fc level.

Number of "up" genes
```{r}
sum(meancounts$log2fc > +2, na.rm=TRUE)
```

> Q9. Determine how many down regulated genes we have at the greater than 2 fc level.

Number of "down" genes at -2 threshold
```{r}
sum(meancounts$log2fc < -2, na.rm=TRUE)
```
> Q10. Do you trust these results? Why or why not?

I do not trust these results. This method of calculating the number of "up" and "down" genes fails to consider possible outliers/values that are not statistically significant by including these values in the results.

## DESeq2 analysis

Let's do this analysis properly and keep our inner stats nerd happy - i.e. are the differences we see between drug and no drug significant given the replicate experiments.

```{r, message=FALSE}
library(DESeq2)
```

For DESeq2 analysis, we need three things

- count values (`countData`)
- metadata tells us about the columns in `countData` (`colData`)
- design of the experiment (i.e. what do you want to compare)

Our first function from DESeq2 will setup the input required for analysis by storing all these three things together.

```{r}
dds <- DESeqDataSetFromMatrix(countData=counts,
                              colData=metadata,
                              design=~dex)
```

The main function in DESeq2 that runs the analysis is called `DESeq()`

```{r}
dds <- DESeq(dds)
```

```{r}
res <- results(dds)
head(res)
```

## Volcano plot
 
This is a common summary result figure from these types of experiments and plots the log2 fold change vs. the adjusted p-value. 

```{r}
plot(res$log2FoldChange, -log(res$padj))
abline(v=c(-2,2), col="red")
abline(h=-log(0.05), col="red")
```

## Save our results

```{r}
write.csv(res, file="my_resutls.csv")
```


## Add gene annotation

To help make sense of our results and communicate them to others, we need to add some more annotations to our main `res` object.

We will use two bioconductor packages to first map IDs to different formats including the classic gene "symbol" gene name.

I will install these with the following commands:
`BiocManager::install("AnnotationDbi")`
`BiocManager::install("org.Hs.eg.db")`

```{r}
library(AnnotationDbi)
library(org.Hs.eg.db)
```

Let's see what is in`org.Hs.eg.db` with the `columns()` function:

```{r}
columns(org.Hs.eg.db)
```

We can translate or "map" IDs between any of these 26 databases using the `mapIds()` function.

```{r}
res$symbol <- mapIds(keys = row.names(res), # our current IDs
                    keytype = "ENSEMBL",    # the format of our IDs
                    x = org.Hs.eg.db,      # where to get the mappings from
                    column =  "SYMBOL")    # the format/DB to map to
head(res)
```

Add the mappings for "GENENAME" and "ENTREZID" and store as `res$genename` and `res$entrez`

```{r}
res$genename <- mapIds(keys = row.names(res),
       keytype = "ENSEMBL",
       x = org.Hs.eg.db,
       column = "GENENAME"
       )
```
```{r}
res$entrez <- mapIds(keys = row.names(res),
       keytype = "ENSEMBL",
       x = org.Hs.eg.db,
       column = "ENTREZID"
       )
```
```{r}
head(res)
```

## Pathway analysis

There are lots of bioconductor packages to do this type of analysis. FOr now, let's just try one called **gage** again we need to install this if we don't have it already.

```{r, message=FALSE}
library(gage)
library(gageData)
library(pathview)
```

To use **gage**, I need two things 

- a named vector of fold-change values for our DEGs (our geneset of interest)
- a set of pathways or geneset to use for annotation

```{r}
x <- c("barry"=5, "lisa"=10)
x
names(x) <- c("low", "high")
x
```

```{r}
foldchanges <- res$log2FoldChange
names(foldchanges) <- res$entrez
head(foldchanges)
```

```{r}
data(kegg.sets.hs)

keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

In our results object we have:
```{r}
attributes(keggres)
```

```{r}
head(keggres$less, 5)
```

Let's look at one of these pathways (hsa05310 Asthma) with our genes colored up so we can see the overlap 

```{r}
pathview(pathway.id = "hsa05310", gene.data = foldchanges)
```

Add this pathway figure to our lab report

![](hsa05310.pathview.png)

## Save our main results

```{r}
write.csv(res, file = "myresults_annotated.csv")
```




